[RemoveDIs] C API: Add before-dbg-record versions of IRBuilder position funcs

OCHyams opened the pull request:
Add `LLVMPositionBuilderBeforeDbgRecords` and
`LLVMPositionBuilderBeforeInstrAndDbgRecords` to `llvm/include/llvm-c/Core.h`
which behave the same as `LLVMPositionBuilder` and `LVMPositionBuilderBefore`
except that the position is set before debug records attached to the target
instruction (the existing functions set the insertion point to after any
attached debug records).

More info on debug records and the migration towards using them can be found
here: https://llvm.org/docs/RemoveDIsDebugInfo.html
 
The distinction is important in some situations. An important example is when
inserting a phi before another instruction which has debug records attached to
it (these come "before" the instruction). Inserting before the instruction but
after the debug records would result in having debug records before a phi,
which is illegal. That results in an assertion failure:

`llvm/lib/IR/Instruction.cpp:166: Assertion '!isa<PHINode>(this) && "Inserting PHI after debug-records!"' failed.`

In llvm (C++) we've added bit to instruction iterators that carries around the
extra information. Adding dedicated functions seemed like the least invasive
and least suprising way to update the C API.

Update llvm/tools/llvm-c-test/debuginfo.c to test this functionality.

Update the OCaml bindings, the migration docs and release notes.

OCHyams committed:
Add BeforeDbgRecords option to C API insertion functions

Inserting a PHI before an instruction with debug records attached results in
the following assert because the PHI needs to be inserted before its debug
records too.

llvm/lib/IR/Instruction.cpp:166: void llvm::Instruction::insertBefore(llvm::BasicBlock&, llvm::iplist_impl<llvm::simple_ilist<llvm::Instruction, llvm::ilist_iterator_bits<true> >, llvm::SymbolTableListTraits<llvm::Instruction, llvm::ilist_iterator_bits<true> > >::iterator): Assertion `!isa<PHINode>(this) && "Inserting PHI after debug-records!"' failed.

OCHyams committed:
remove other docs updates

alan-j-hu commented:
Just ran the OCaml binding unit tests, they all pass.

alan-j-hu commented:
I just want to fix some lines that go past 80 columns in `llvm.ml`. Although the file already contains a few lines that do go past 80 columns, most of the file is formatted with newlines inserted to avoid this.

alan-j-hu commented on llvm/bindings/ocaml/llvm/llvm.ml:
```diff
@@ -1137,22 +1137,22 @@ external delete_instruction : llvalue -> unit = "llvm_delete_instruction"
 
-let builder_before context i = builder_at context (Before i)
-let builder_at_end context bb = builder_at context (At_end bb)
+let builder_before context i before_dbg_records = builder_at context (Before i) before_dbg_records
+let builder_at_end context bb = builder_at context (At_end bb) false
```
```suggestion
let builder_before context i before_dbg_records =
  builder_at context (Before i) before_dbg_records
let builder_at_end context bb =
  builder_at context (At_end bb) false
```

alan-j-hu commented on llvm/bindings/ocaml/llvm/llvm.ml:
```diff
@@ -1137,22 +1137,22 @@ external delete_instruction : llvalue -> unit = "llvm_delete_instruction"
 external builder : llcontext -> llbuilder = "llvm_builder"
-external position_builder : (llbasicblock, llvalue) llpos -> llbuilder -> unit
-                          = "llvm_position_builder"
+external position_builder : (llbasicblock, llvalue) llpos -> bool -> llbuilder ->
+                            unit = "llvm_position_builder"
```
```suggestion
external position_builder : (llbasicblock, llvalue) llpos -> bool ->
                            llbuilder -> unit
                          = "llvm_position_builder"
```

alan-j-hu commented on llvm/bindings/ocaml/llvm/llvm.ml:
```diff
@@ -1137,22 +1137,22 @@ external delete_instruction : llvalue -> unit = "llvm_delete_instruction"
 
-let position_before i = position_builder (Before i)
-let position_at_end bb = position_builder (At_end bb)
+let position_before i before_dbg_records = position_builder (Before i) before_dbg_records
```
```suggestion
let position_before i before_dbg_records =
  position_builder (Before i) before_dbg_records
```

alan-j-hu reviewed (changes requested):
Add newline to limit to 80 columns

nikic reviewed (changes requested):
It is not possible to change the signature of a function in the C API. You can only add a new function with the new signature.

The usual pattern would be to add LLVMPositionBuilder2 (in this case an alternative would be to add a new function with the parameter set to the non-default value).

OCHyams commented:
Thanks for taking a look @nikic 

> It is not possible to change the signature of a function in the C API. You can only add a new function with the new signature.

Oh, I can't see this rule in the [developer policy](https://llvm.org/docs/DeveloperPolicy.html#c-api-changes). But that would make the changes more stable.

> The usual pattern would be to add LLVMPositionBuilder2 (in this case an alternative would be to add a new function with the parameter set to the non-default value).

The reason we need to modify the API is that at the moment it's possible to insert PHIs after debug records which is illegal and leads to an assertion. I chose to modify existing functions because I figured would help reduce nasty surprises, but if stability is preferable this isn't the right tactic.

We could add new functions with the extra parameter, keep the existing functions untouched and forward args and a "default" value for the new parameter to the new functions (possibly this is what you're suggesting already)? I guess this would be most "stable" (and I think then we can omit most of the OCaml changes then).

I wonder if some of the other changes we made need re-reviewing in this case:
https://github.com/llvm/llvm-project/pull/84915 - adds some new functions (I imagine this patch is fine)
https://github.com/llvm/llvm-project/pull/85657 - changes return value of existing function (is this one ok?)

nikic commented:
> Thanks for taking a look @nikic
> 
> > It is not possible to change the signature of a function in the C API. You can only add a new function with the new signature.
> 
> Oh, I can't see this rule in the [developer policy](https://llvm.org/docs/DeveloperPolicy.html#c-api-changes). But that would make the changes more stable.

The phrasing here as a bit unclear as to what "best effort stability" means, but the rules are basically:

 * You are allowed to add new functions.
 * You shouldn't remove functions, but are allowed to do so if there's a strong reason (e.g. the underlying feature has been removed from LLVM).
 * You are never allowed to make an ABI-breaking change to an existing function.

Changes to the C API should, at worst, produce a linker error, but never introduce an FFI mismatch.

> > The usual pattern would be to add LLVMPositionBuilder2 (in this case an alternative would be to add a new function with the parameter set to the non-default value).
> 
> The reason we need to modify the API is that at the moment it's possible to insert PHIs after debug records which is illegal and leads to an assertion (edit: more specifically, there's no way to _avoid_ that behaviour using the C API with the new debug info format - inserting a phi before an instruction with debug records attached will always cause the assertion without this patch). I chose to modify existing functions because I figured would help reduce nasty surprises, but if stability is preferable this isn't the right tactic.

I think this does not affect typical use of the C API (where a frontend emits instructions in order -- it will not go back to a previous position), so I wouldn't be overly concerned about accidental misuse.
 
> We could add new functions with the extra parameter, keep the existing functions untouched and forward args and a "default" value for the new parameter to the new functions (possibly this is what you're suggesting already)? I guess this would be most "stable" (and I think then we can omit most of the OCaml changes then).

Yes, that's what I would suggest.

> I wonder if some of the other changes we made need re-reviewing in this case: #84915 - adds some new functions (I imagine this patch is fine)

This is fine.

> #85657 - changes return type of existing functions (is this one ok?)

Yeah, this is not allowed :(

I think what I'd do for the C API is to keep just the new functions you added for dbg records and remove the old functions for intrinsics. Keeping the old name around and making it create intrinsics would be too much of a footgun. And I don't think that we actually need support for the old format in the C API.

OCHyams commented:
Ok thanks for the clarifications and explanations. I'll make those changes to this patch and then follow up with a patch to fix my previous changes too.

> Yeah, this is not allowed :(
> I think what I'd do for the C API is to keep just the new functions you added for dbg records and remove the old functions for intrinsics. Keeping the old name around and making it create intrinsics would be too much of a footgun. And I don't think that we actually need support for the old format in the C API.

For the intrinsic-inserting functions, do you think we should:
a) mark as deprecated but otherwise leave them alone for now
b) mark as deprecated and change their behaviour so that they do nothing and return nullptr
c) outright delete them (it sounds like you're voting this one, but just wanted to double check)

OCHyams committed:
add new functions rather than update existing

OCHyams committed:
undo ocaml changes

OCHyams committed:
update llvm-c-test/debuginfo.c

OCHyams committed:
update ocaml bindings

OCHyams committed:
typo

OCHyams commented:
>> We could add new functions with the extra parameter, keep the existing functions untouched and forward args and a "default" value for the new parameter to the new functions (possibly this is what you're suggesting already)? I guess this would be most "stable" (and I think then we can omit most of the OCaml changes then).

> Yes, that's what I would suggest.

Done, how does this look now?

@alan-j-hu sorry for the turbulence but could you re-review the OCaml changes? They're much smaller this time.

alan-j-hu reviewed (changes requested):
First of all, you wrote a second declaration for `position_before` in `llvm.mli` where I assume you meant to declare `position_before_dbg_records`. Consequently, the tests do not pass.

Are `LLVMPositionBuilder` and `LLVMPositionBuilder2` intended to co-exist forever, or will the latter eventually deprecate the former?

If `LLVMPositionBuilder` will be deprecated, I would rather change `position_builder` right now to take the `bool`, and call the underlying C function `LLVMPositionBuilder2` than do the song and dance with the `2`. As far as I know, OCaml libraries don't have these kinds of ABI compatibility concerns, it's much better to rip off the band-aid once, between two major releases, than go through multiple steps of deprecation across multiple major releases.

Are `position_before` and `position_before_dbg_records` intended to co-exist, or will the latter eventually deprecate the former? By inserting after debug records, does `position_before` have "surprising" behavior that may cause errors?

If `position_before` has "surprising" behavior, or behavior that shouldn't be the "default," I would rather have it behave as `position_before_dbg_records`, i.e. inserting before debug records be the default action unless the programmer says otherwise.

I'm not an expert on the debug record migration, so please let me know if:

- If I didn't explain any of my suggestions clearly
- Any of my suggestions don't make sense
- Any of my suggestions may cause issues for users

alan-j-hu commented on llvm/bindings/ocaml/llvm/llvm_ocaml.c:
```diff
@@ -2016,18 +2016,28 @@ value llvm_builder(value C) {
+  return llvm_position_builder_impl(Pos, B, Bool_val(BeforeDbgRecords));
+}
+
+/* (llbasicblock, llvalue) llpos -> llbuilder -> unit */
+value llvm_position_builder(value Pos, value B) {
+  return llvm_position_builder_impl(Pos, B, /*BeforeDbgRecords=*/0);
+}
```
Whoops, somehow this didn't get attached in my code review.

I would very much rather implement this in `llvm.ml` as:

```ocaml
let position_builder llpos llbuilder = position_builder2 llpos false llbuilder
```

That being said, I am questioning whether `position_builder` and `position_buidler2` should co-exist. OCaml doesn't have the ABI compatibility concerns between major library releases, so it's likely much better to rip off the bandaid with a single change to the type of `position_builder`, instead of going through the process of with the `2`.

alan-j-hu commented on llvm/bindings/ocaml/llvm/llvm.ml:
```diff
@@ -1152,6 +1154,7 @@ let builder_before context i = builder_at context (Before i)
 let builder_at_end context bb = builder_at context (At_end bb)
 
 let position_before i = position_builder (Before i)
+let position_before_dbg_records i = position_builder2 (Before i) true
```
Whoops, this didn't get attached to my code review.

I am wondering about whether positioning after the debug records is the correct "default" behavior.

Would an optional argument be an appropriate choice here?

In `llvm.mli`:

```ocaml
val position_before : ?before_dbg_records:bool -> llpos -> llbuilder -> unit
```

In `llvm.ml`:

```ocaml
let position_builder (?before_dbg_records=true) i = position_builder2 (Before i) before_dbg_records
```

OCHyams replied:
I've side-stepped this in the rework that I've just uploaded - now there are just two separate functions for inserting after/before debug records - there's no bool parameter flying around.

alan-j-hu reviewed (changes requested):
Attaching comments that I forgot to explicitly add to the code review.

OCHyams commented:
> First of all, you wrote a second declaration for `position_before` in `llvm.mli` where I assume you meant to declare `position_before_dbg_records`. Consequently, the tests do not pass.

Oops, I'll fix that.

> Are `LLVMPositionBuilder` and `LLVMPositionBuilder2` intended to co-exist forever, or will the latter eventually deprecate the former?

With the latest changes - intended to co-exist forever.

> If `LLVMPositionBuilder` will be deprecated, I would rather change `position_builder` right now to take the `bool`, and call the underlying C function `LLVMPositionBuilder2` than do the song and dance with the `2`. As far as I know, OCaml libraries don't have these kinds of ABI compatibility concerns, it's much better to rip off the band-aid once, between two major releases, than go through multiple steps of deprecation across multiple major releases.

> Are `position_before` and `position_before_dbg_records` intended to co-exist, or will the latter eventually deprecate the former? By inserting after debug records, does `position_before` have "surprising" behavior that may cause errors?

There is some room for surprising behaviour imo which is why I took my initial approach, however, see @nikic's [comment](https://github.com/llvm/llvm-project/pull/92417#issuecomment-2120078326) - typical api usage probably won't run into it.

Assuming a front end emits instructions in order, most of the time it'll just be inserting at the end of a block or before a terminator. In both these scenarios there's no surprising behaviour and `position_before` does the right thing. It would actually be a mistake to use `position_before_dbg_records` in the case of inserting in order before a terminator, because then all the debug records would end up collecting above the terminator rather than in order of insertion.

If a tool using this api inserts instructions in reverse order by repeatedly setting the "insert before" position then `position_before` would be incorrect - all the debug records would get pushed up to the top of the block. There are other situations where `position_before` is wrong too, which I've described in other comments (e.g. inserting phis).

> If `position_before` has "surprising" behavior, or behavior that shouldn't be the "default," I would rather have it behave as `position_before_dbg_records`, i.e. inserting before debug records be the default action unless the programmer says otherwise.

The `position_before` behaviour is closer to the default in LLVM (LLVM uses a different mechanism to determine "before dbg records", a flag in instruction iterators, and most of the time that is false).

---

I'll wait for @nikic to comment before making further changes - or is it worth opening a discourse discussion at this point?

alan-j-hu commented:
- If `LLVMPositionBuilder` and `LLVMPositionBuilder2` are intended to co-exist - i.e., the latter is not intended to replace the former in any future release, I'm wondering if `LLVMPositionBuilder2` should have a more descriptive name. (The main case where I've seen the usage of `2` is the migration from typed to opaque pointers, where the old functions were eventually removed. Have there been any instances where a `2` function was added that didn't deprecate the older function?) For both the C and OCaml APIs, just adding a `2` at the end of the name doesn't indicate to the user what this function does differently from the other function.

- I would rather see `position_builder` implemented in OCaml code, not C FFI code. i.e.

  ```ocaml
  let position_builder ip = position_builder2 ip false
  ```

- Don't forget to replace the errorneous second declaration of `position_before` with the intended declaration of `position_before_dbg_records` in `llvm.mli`.

These are my only comments now; you have justified your design choices about `position_before` and `position_before_dbg_records`.

OCHyams commented:
Ping @nikic wdyt of the discussion above?

@alan-j-hu 
> If LLVMPositionBuilder and LLVMPositionBuilder2 are intended to co-exist - i.e., the latter is not intended to replace the former in any future release, I'm wondering if LLVMPositionBuilder2 should have a more descriptive name. [...]

I think that makes sense but I'm struggling to think of a suitable name (any ideas?). An alternative could be to drop the bool argument from the new functions and add a "BeforeDbgRecords" suffix to the names (though this leaves with with `PositionBuilderBeforeDbgRecords` and `PositionBuilderBeforeBeforeDbgRecords`, the latter of which is ugly and looks like a typo). @nikic do you have an opinion on using `2` in the name for functions that are not getting deprecated?

nikic commented:
> For the intrinsic-inserting functions, do you think we should: a) mark as deprecated but otherwise leave them alone for now b) mark as deprecated and change their behaviour so that they do nothing and return nullptr c) outright delete them (it sounds like you're voting this one, but just wanted to double check)

I'd go for c), yes.

> * If `LLVMPositionBuilder` and `LLVMPositionBuilder2` are intended to co-exist - i.e., the latter is not intended to replace the former in any future release, I'm wondering if `LLVMPositionBuilder2` should have a more descriptive name. (The main case where I've seen the usage of `2` is the migration from typed to opaque pointers, where the old functions were eventually removed. Have there been any instances where a `2` function was added that didn't deprecate the older function?) For both the C and OCaml APIs, just adding a `2` at the end of the name doesn't indicate to the user what this function does differently from the other function.

You are right that `2` usually indicates that the old function is deprecated -- though in some cases it is soft-deprecated without intent to remove (e.g. when we're just changing parameter types from unsigned to uint64_t or similar).

The naming pattern that often gets used for cases like this is "With", so we have both LLVMBuildCall2 and LLVMBuildCallWithOperandBundles, where the latter just provides additional functionality, but most use-cases can still use the former just fine. This is similar to the situation here.

So I guess you could call it LLVMPositionBuilderWithDbgRecordControl, which is, uh, quite the mouthful :) I don't really have strong opinions on the precise naming here.

OCHyams committed:
rework patch

* C: add new versions without bool param
* ocaml: position_before_dbg_records stays the same (equ. to LLVMPositionBuilderBeforeInstrAndDbgRecords)
* ocaml: position_builder2 -> position_builder_before_dbg_records (equ. to LLVMPositionBuilderBeforeDbgRecords)

OCHyams commented:
Would either of you be opposed to the new functions unconditionally inserting before debug records instead of providing the option through a bool parameter (given you can choose the "insert after debug records" behaviour by calling the existing functions which we don't plan to deprecate)?

OCHyams committed:
update C API comments

OCHyams commented:
Ok I've reworked the patch:

* C: Added `LLVMPositionBuilderBeforeDbgRecords` and `LLVMPositionBuilderBeforeInstrAndDbgRecords`. Same as `LLVMPositionBuilder` and `LLVMPositionBuilderBefore` except the insertion position is set to before the debug records that precede the target instruction (the bool parameter has been removed from the new functions).
* OCaml: Added `position_builder_before_dbg_records` which does the same thing as `LLVMPositionBuilderBeforeDbgRecords` and `position_before_dbg_records` which does the same thing as `LLVMPositionBuilderBeforeInstrAndDbgRecords`. The OCaml names are different to and more sensible than the C ones because  unlike the C ones the existing OCaml function names lend themselves to simply appending "before_dbg_records".

nikic commented on llvm/include/llvm-c/Core.h:
```diff
@@ -3957,9 +3957,28 @@ const unsigned *LLVMGetIndices(LLVMValueRef Inst);
 LLVMBuilderRef LLVMCreateBuilderInContext(LLVMContextRef C);
 LLVMBuilderRef LLVMCreateBuilder(void);
+/**
+ * Set the builder poisiton before Instr but after any attached debug records,
```
```suggestion
 * Set the builder position before Instr but after any attached debug records,
```
Same below.

nikic reviewed (approved):
The C API changes look fine to me. Would be good if @alan-j-hu takes a look at the new OCaml changes.

OCHyams committed:
fix typo

OCHyams commented:
Rebase + fix typo. I'll land this shortly.

Thanks both for your patience and reviews.
